### PR TITLE
[rl] [1/n] Refactor policy trainer

### PR DESCRIPTION
**Pull request by joecummings**
Goals of this refactor:

Simplify, do less.

Manifests as:
1. Match the existing patterns of Torchtitan's trainer for the pieces we use. If the same things look different in two places, it's just confusing
2. Solidify API and introduce data types. We borrow ideas from forge and tinker on how a training client should be set up. We introduce some datatypes that are **very flexible** just to show that we should be fitting clean boundaries here

Everything else is downstream of these changes.

### CODE REVIEW

File: torchtitan/experiments/rl/actors/trainer.py [hunk 1]
```diff
@@ -59,184 +48,192 @@ class Config(Configurable.Config):
...
-        # Build optimizer and LR scheduler
+        # Build optimizer and LR scheduler (after parallelize, matches main Trainer)
         self.optimizers = config.optimizer.build(model_parts=self.model_parts)
         self.lr_schedulers = config.lr_scheduler.build(
             optimizers=self.optimizers,
             training_steps=config.training.steps,
         )
 
+        # RL-specific state
         self.policy_version = 0
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:120]: nit: let's put this towards the end of this function

File: torchtitan/experiments/rl/actors/trainer.py [hunk 2]
```diff
@@ -6,51 +6,40 @@
...
-
-    Uses ModelSpec for model construction, parallelization, and weight loading.
-
-    Args:
-        config: PolicyTrainer.Config for model/optimizer/parallelism settings.
-        model_spec: Model specification (model config, parallelize_fn, state_dict_adapter).
-        hf_assets_path: Path to HF assets folder for checkpoint loading.
-    """
-
+class PolicyTrainer(Actor, Stateful, Configurable):
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:35]: should this just inherit main trainer (and actor)?

They don't share the same set of functions, but we may be able to reduce the constant effort to update both sides. We may explicit inherit and ban the functions we don't need here.

File: torchtitan/experiments/rl/actors/trainer.py [hunk 3]
```diff
@@ -59,184 +48,192 @@ class Config(Configurable.Config):
...
         parallelism: ParallelismConfig = field(default_factory=ParallelismConfig)
-        comm: CommConfig = field(default_factory=CommConfig)
-        """Communication configuration for distributed initialization."""
         compile: CompileConfig = field(default_factory=CompileConfig)
+        comm: CommConfig = field(default_factory=CommConfig)
+        checkpoint: CheckpointManager.Config = field(
+            default_factory=CheckpointManager.Config
+        )
+
+    weight_sync_transfer_dtype: Literal["bfloat16", "float32"] = "bfloat16"
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:57]: this seems hardcoded, should it be put in config?

File: torchtitan/experiments/rl/actors/trainer.py [hunk 4]
```diff
@@ -59,184 +48,192 @@ class Config(Configurable.Config):
...
 
-        return model
+        return OptimStepResult(
+            grad_norm=(grad_norm.item() if hasattr(grad_norm, "item") else grad_norm),
+            policy_version=self.policy_version,
+        )
 
     @endpoint
-    async def push_model_state_dict(self) -> None:
+    async def push_weights(self) -> None:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:256]: I had the comment that "weights" doesn't sound formal -- if we use weights, what about "bias"?

File: torchtitan/experiments/rl/actors/trainer.py [hunk 5]
```diff
@@ -254,126 +251,16 @@ async def push_model_state_dict(self) -> None:
...
-                    dist.all_reduce(param.grad, op=dist.ReduceOp.AVG)
-
-        # Gradient clipping
-        grad_norm = dist_utils.clip_grad_norm_(
-            [p for m in self.model_parts for p in m.parameters()],
-            self.config.training.max_norm,
-            foreach=True,
-            pp_mesh=self.parallel_dims.get_optional_mesh("pp"),
+            transfer_dtype=TORCH_DTYPE_MAP[self.weight_sync_transfer_dtype]
+            if self._cast_dtypes_for_weight_sync
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:277]: why do we need this variable?

File: torchtitan/experiments/rl/actors/trainer.py (hunk 2 above)
Comment (joecummings) [reply torchtitan/experiments/rl/actors/trainer.py:35]: I'd like to move in that direction, yes - this is the first step. That would require some updates to the main trainer so starting by just making them as similar as possible. That way it's easier for an LLM to pick up changes and apply regardless.

Comment (tianyu-l) [reply torchtitan/experiments/rl/actors/trainer.py:35]: Sg. Just fyi we are still doing refactor to main trainer. @tugsbayasgalan and I will be adding eager chunked ce loss, which will necessitate changing trainer a bit.
